### PR TITLE
Update to ports polling

### PR DESCRIPTION
**Pull request by laf**
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

I'll put inline comments in place.

---
This change is [Reviewable](https://reviewable.io/reviews/librenms/librenms/7080)

### CODE REVIEW

File: includes/polling/ports.inc.php [hunk 1]
```diff
@@ -66,20 +66,6 @@
     'ifOutMulticastPkts',
 );
 
-$cisco_oids = array(
```
Comment (laf) [inline includes/polling/ports.inc.php]: Not used at all

File: includes/polling/ports.inc.php [hunk 2]
```diff
@@ -201,6 +188,7 @@
                         $port_stats[$i] = $data[$i];
                     }
                     $port_stats = snmp_get_multi($device, $oids, '-OQUst', 'IF-MIB', null, $port_stats);
+                    $port_stats[$i]['polled'] = time();
```
Comment (laf) [inline includes/polling/ports.inc.php]: Before hand, we obtained the polled time for the ports way after all the snmpwalks when we actually only want to do it after ifEntry/ifXEntry. So we now set this either on a per port basis for selected ports polling or right after the if(X)Entry runs so we get as close to accurate time as possible.

File: includes/polling/ports.inc.php [hunk 3]
```diff
@@ -472,14 +460,22 @@
...
-        // use HC values if they are available
-        foreach ($hc_mappings as $hc_oid => $if_oid) {
-            if (isset($this_port[$hc_oid]) && $this_port[$hc_oid]) {
-                d_echo("$hc_oid ");
-                $this_port[$if_oid] = $this_port[$hc_oid];
-            } else {
-                d_echo("$if_oid ");
+        $hc_enabled = false;
+        // use HC values if they are available and not disabled
+        if ($config['os'][$device['os']]['no_hc'] != true) {
```
Comment (laf) [inline includes/polling/ports.inc.php]: Added the option to be able to skip using hc counters. I've helped someone recently where the hc values are utterly broken.

File: includes/polling/ports.inc.php [hunk 4]
```diff
@@ -641,6 +631,20 @@
 
             $oid_prev = $oid.'_prev';
             if (isset($port[$oid])) {
+                if ($device['uptime'] < $config['rrd']['step']) {
```
Comment (laf) [inline includes/polling/ports.inc.php]: This block attempts to address counter rollover for ports in 32 bit mode.

If the device hasn't recently rebooted and the current Octet value is < then the previous something has gone wrong. So we subtract the previous Octet value from the max possible for 64bit and then add the current one onto that value.

**Comment by laf** [comment]
I might switch the rolled calculation to this one: https://blog.chasefox.net/calculating-ifinoctets/

Anyone got any comments on that?

**Comment by kkrumm1** [comment]
is this only for cisco devices? or all devices?

**Comment by laf** [comment]
@mathias-combelle It's an option per OS, so you can set:

`$config['os']['merakimx']['no_hc'] = true;` in config.php. Feel free to test this pull request and report back.

**Comment by laf** [comment]
@kkrumm1 the general changes are for all devices. The one to skip using HC counters is user configurable per OS.

**Comment by mathias-combelle** [comment]
@laf : Is it in the update or should I test it manually ?
I'm on DEV auto-update mode and do not see the new code.

Thanks,

**Comment by murrant** [comment]
@mathias-combelle This is code that is being reviewed for inclusion.

To test it, follow the instructions at the top of the ticket:
```
If you would like to test this pull request then please run: ./scripts/github-apply <pr_id>, i.e ./scripts/github-apply 7080
```

**Comment by raphbap** [comment]
Hi, I tried it but unfortunately it stopped the polling of Meraki devices (got the notifications + no more data).

When I disabled the option in config.php with this command: 
`$config['os']['merakimx']['no_hc'] = false;`, I kept getting the following pop-up message (although Meraki devices were polled again): 

> It appears as though you have some devices that haven't completed polling within the last 15 minutes, you may want to check that out :)

**Comment by laf** [comment]
please post the output on pastebin of `./poller.php -h HOSTNAME -d -m ports`

**Comment by raphbap** [comment]
Here it is :) : https://pastebin.com/QQ7gKN9F

**Comment by laf** [comment]
Remove the patch, you don't have a module installed (which I thought was part of php core). `./scripts/github-remove -d`

**Comment by laf** [comment]
Updated to remove dependancy on gmp_sub()

File: includes/polling/ports.inc.php [hunk 5]
```diff
@@ -641,6 +631,25 @@
...
+                        $max_64 = '18446744073709551616';//Unused at the moment.
+                        $max_32 = '4294967296';
+                        if ($hc_enabled !== true) {
+                            if (function_exists('gmp_sub')) {
+                                // We store the new rolled version for use in rrd. DO NOT store in the DB as we need
+                                // to calculate the value again on the next poll.
+                                $this_port[$oid . '_rolled'] = (gmp_sub($max_32, $port[$oid])) + $this_port[$oid];
+                            } else {
+                                // No GMP http://uk1.php.net/manual/en/book.gmp.php function
+                                $this_port[$oid . '_rolled'] = ($max_32 - $port[$oid]) - $this_port[$oid];
```
Comment (murrant) [inline includes/polling/ports.inc.php]: Shouldn't the second `-` be `+`?

File: includes/polling/ports.inc.php (hunk 3 above)
Comment (murrant) [inline includes/polling/ports.inc.php]: You could use the new config class here:
`Config::getOsSetting($device, 'no_hc', false)`

**Comment by murrant** [comment]
I thought RRD automatically handles rolled over counters.

**Comment by laf** [comment]
As far as I'm aware no, if you push more than 125Mbps on v1 snmp (or non-HC counters on any version) then you get a flat capped graph or drops in data.

File: includes/polling/ports.inc.php (hunk 5 above)
Comment (laf) [inline includes/polling/ports.inc.php]: Corrected.

I actually had 64Bit in there originally as well when testing for another fix but it produced crazy numbers. I think that was due to the device itself so we could possibly add rollover for 64Bit and at which point GMP is better for precision.

I can drop for now if you like.

**Comment by laf** [comment]
bump

**Comment by mathias-combelle** [comment]
I tried : 
/scripts/github-apply 7080
But got the same issue (looks like it still search for this missing gmp_sub() module).

Then, after a "./scripts/github-remove -d" I tried with : 
./scripts/github-apply 5926
But got an error applying the patch :/

"
error: patch failed: scripts/github-remove:**_xx_**
error: scripts/github-remove: patch does not apply
"
I will wait for a new update

**Comment by laf** [comment]
> But got the same issue (looks like it still search for this missing gmp_sub() module).

`if (function_exists('gmp_sub')) {`

Nope, we check first. Seems like you've not managed to re-apply the patch. Use github-remove then do `git status` to make sure your branch is clean.

**Comment by mathias-combelle** [comment]
"On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working tree clean"

Then /github-apply 5926 -> Fail

**Comment by laf** [comment]
The branch is clean against the current master so either your install is too old or another reason.

You could just download the file and copy it over.

**Comment by mathias-combelle** [comment]
Just run validate.php, made some changes and force the update. Now it has applied the changes for sure :)

I try again and let you know, thanks for your support and tips

**Comment by laf** [comment]
@librenms/reviewers Any comments to this or I'll be merging in in 48 hours.

**Comment by murrant** [comment]
The code itself looks good.  I am just unsure about the usage of gmp.  But not enough to not merge, so +1

**Comment by Rosiak** [comment]
I'll see if I can test this tomorrow.

**Comment by laf** [comment]
Don't bother for now @Rosiak.

**Comment by mathias-bap-git** [comment]
Hi veryone, has this feature been pushed ?
It's still indicating "Merge in 48hours" for a while. 
I still have gaps in all my graphs (Meraki...) so I was really looking forward to this solution.

Thanks in advance,

**Comment by f0o** [comment]
It's been closed rather than merged, chances are that there was a blocker somewhere.

I haven't followed this PR, best to ask laf directly in discord or IRC :)

**Comment by mathias-bap-git** [comment]
Thanks, that's what I was affraid of. 
I will ask him on IRC, thanks for the reply !

**Comment by murrant** [comment]
@mathias-bap-git I thought your issue was 32bit counters on the Meraki, this is for 64bit counter overflows, which would not help for 32bit counters.

**Pull request closed without merging**